pkg/localstore/reserve_test.go

```diff
@@ -0,0 +1,278 @@
+
+// TestDB_ReserveGC_AllWithinRadius tests that when all chunks fall within
+// batch radius, none get collected.
+func TestDB_ReserveGC_AllWithinRadius(t *testing.T) {
+	chunkCount := 150
+
+	var closed chan struct{}
+	testHookCollectGarbageChan := make(chan uint64)
+	t.Cleanup(setTestHookCollectGarbage(func(collectedCount uint64) {
+		select {
+		case testHookCollectGarbageChan <- collectedCount:
+		case <-closed:
+		}
+	}))
+
+	db := newTestDB(t, &Options{
+		Capacity: 100,
+	})
+	closed = db.close
+
+	addrs := make([]swarm.Address, 0)
+
+	for i := 0; i < chunkCount; i++ {
+		ch := generateTestRandomChunkAt(swarm.NewAddress(db.baseKey), 2).WithBatch(2, 3)
+		err := db.UnreserveBatch(ch.Stamp().BatchID(), 2)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = db.Put(context.Background(), storage.ModePutUpload, ch)
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = db.Set(context.Background(), storage.ModeSetSync, ch.Address())
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		addrs = append(addrs, ch.Address())
+	}
+
+	select {
+	case <-testHookCollectGarbageChan:
+		t.Fatal("gc ran but shouldnt have")
+	case <-time.After(1 * time.Second):
+	}
+
+	t.Run("pull index count", newItemsCountTest(db.pullIndex, chunkCount))
+
+	t.Run("postage chunks index count", newItemsCountTest(db.postageChunksIndex, chunkCount))
+
+	t.Run("gc index count", newItemsCountTest(db.gcIndex, 0))
+
+	t.Run("gc size", newIndexGCSizeTest(db))
+
+	t.Run("all chunks should be accessible", func(t *testing.T) {
+		for _, a := range addrs {
+			_, err := db.Get(context.Background(), storage.ModeGetRequest, a)
+			if err != nil {
+				t.Errorf("got error %v, want none", err)
+			}
+		}
+	})
+}
+
+// TestDB_ReserveGC_Unreserve tests that after calling UnreserveBatch
+// with a certain radius change, the correct chunks get put into the
+// GC index and eventually get garbage collected.
+// batch radius, none get collected.
+func TestDB_ReserveGC_Unreserve(t *testing.T) {
+	chunkCount := 150
+
+	var closed chan struct{}
+	testHookCollectGarbageChan := make(chan uint64)
+	t.Cleanup(setTestHookCollectGarbage(func(collectedCount uint64) {
+		select {
+		case testHookCollectGarbageChan <- collectedCount:
+		case <-closed:
+		}
+	}))
+
+	db := newTestDB(t, &Options{
+		Capacity: 100,
+	})
+	closed = db.close
+
+	// put the first chunkCount chunks within radius
+	for i := 0; i < chunkCount; i++ {
+		ch := generateTestRandomChunkAt(swarm.NewAddress(db.baseKey), 2).WithBatch(2, 3)
+		err := db.UnreserveBatch(ch.Stamp().BatchID(), 2)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = db.Put(context.Background(), storage.ModePutUpload, ch)
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = db.Set(context.Background(), storage.ModeSetSync, ch.Address())
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	var gcChs []swarm.Chunk
+	for i := 0; i < 100; i++ {
+		gcch := generateTestRandomChunkAt(swarm.NewAddress(db.baseKey), 2).WithBatch(2, 3)
+		err := db.UnreserveBatch(gcch.Stamp().BatchID(), 2)
```

Comment:
maybe a better test would be to add a few chunks of at PO 4 and show they are intact.
otherwise this is a very good test`